### PR TITLE
Onyx Migration for checkPreviousReportActionID

### PR DESCRIPTION
**Pull request by neonbhai**
### Details
Adds the Onyx Migration for checkPreviousReportActionID

### Fixed Issues

$ https://github.com/Expensify/App/issues/23221
PROPOSAL: https://github.com/Expensify/App/issues/23221#issuecomment-1642956270

### Tests

Automated Test Steps:
1. Run `npx jest MigrationTest.js`
2. Verify all tests pass for CheckForPreviousReportActionID

Manual Test Steps
1. Add CheckForPreviousReportActionID in [this file](https://github.com/Expensify/App/blob/218920473e5753234408554cfd86517e4923ad84/src/libs/migrateOnyx.js#L19-L26)
1. Open Expensify.
3. If logged out: No Report Actions so Migration skipped. 
4. See console message: `[Migrate Onyx] Skipped migration CheckForPreviousReportActionID because there were no reportActions`
5. If logged in: The migration will be carried out: Since currently no report actions have the previousReportActionID field. The migration will delete all reportActions from onyx.
6. See console message: `[Migrate Onyx] CheckForPreviousReportActionID Migration: removing all reportActions because previousReportActionID not found in the first reportAction`

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>
# Tests:

image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by abdulrahuman5196** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>






</details>

<details>
<summary>Mobile Web - Chrome</summary>






</details>

<details>
<summary>Mobile Web - Safari</summary>







</details>

<details>
<summary>Desktop</summary>







</details>

<details>
<summary>iOS</summary>






</details>

<details>
<summary>Android</summary>






</details>

File: tests/unit/MigrationTest.js [hunk 1]
```diff
@@ -639,4 +640,72 @@ describe('Migrations', () => {
                     });
                 }));
     });
+
+    describe('CheckPreviousReportActionID', () => {
+        // Note: this test has to come before the others in this suite because Onyx.clear leaves traces and keys with null values aren't cleared out between tests
```
Comment (abdulrahuman5196) [inline tests/unit/MigrationTest.js:645]: Why is this? I tried moving the tests and there was no change and all tests passed.

File: tests/unit/MigrationTest.js [hunk 2]
```diff
@@ -639,4 +640,72 @@ describe('Migrations', () => {
                     });
                 }));
     });
+
+    describe('CheckPreviousReportActionID', () => {
```
Comment (abdulrahuman5196) [inline tests/unit/MigrationTest.js:644]: The tests are very limited to only solve the happy case. Can we add some corner case tests like
{reportAction1: null} or {reportAction1: {}} ?

**Review by abdulrahuman5196 — CHANGES_REQUESTED** [review]

**Comment by abdulrahuman5196** [comment]
@neonbhai Kindly check on the above comments

File: tests/unit/MigrationTest.js (hunk 2 above)
Comment (neonbhai) [reply tests/unit/MigrationTest.js:644]: Okay, adding. What would be the scenario in which a `reportAction` is null or an empty object? Since an 'action' object is typically expected to have content, I am guessing that `reportActions` are not in the right condition. Should the check for `previousReportActionID` then fail (and proceed with removing all report actions)? As the migration did not find the field.

File: src/libs/migrations/CheckForPreviousReportActionID.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ */
+export default function () {
+    return getReportActionsFromOnyx().then((allReportActions) => {
+        if (_.isEmpty(allReportActions)) {
+            Log.info(`[Migrate Onyx] Skipped migration CheckForPreviousReportActionID because there were no reportActions`);
+            return;
+        }
+
+        const firstReportActionID = _.keys(allReportActions)[0];
+        const firstReportAction = allReportActions[firstReportActionID];
```
[the remark this answers is not in the document]
Comment (neonbhai) [reply src/libs/migrations/CheckForPreviousReportActionID.js:36]: since reportActions was a complex object, I spelled out how I got the first value of first reportActionID in accordance to clean code. 
`const firstValueOfFirstReportAction = _.values(_.values(allReportActions)[0])[0];`
This did look a bit cryptic to me at first. Should I change it to 1 line?

File: tests/unit/MigrationTest.js (hunk 1 above)
Comment (neonbhai) [reply tests/unit/MigrationTest.js:645]: Hi I saw this note in a previous test [here](https://github.com/Expensify/App/blob/975146fa30bc65a19df8e0cca525af32ce2ceda5/tests/unit/MigrationTest.js#L153). Reordering was also passing for me, but I was not sure if this was an edge case in onyx and doesn't happen everytime. So added for clarity. Should I remove it?

File: tests/unit/MigrationTest.js (hunk 2 above)
Comment (abdulrahuman5196) [reply tests/unit/MigrationTest.js:644]: @roryabraham What should we do in case of zoombie reportActions? like {reportAction1: null} or {reportAction1: {}}.
Should we loop through to find other valid reportActions and check previousReportActionID or just clear all?

Comment (roryabraham) [reply tests/unit/MigrationTest.js:644]: I agree we should have some edge case handling for zombie data and the like, just so that this code is nice and robust.

> What should we do in case of zoombie reportActions? like {reportAction1: null} or {reportAction1: {}}.
Should we loop through to find other valid reportActions and check previousReportActionID or just clear all?

I would suggest that we loop through until a reportAction w/ a `reportActionID` field is found, then proceed with the existing logic.

Comment (abdulrahuman5196) [reply tests/unit/MigrationTest.js:644]: @neonbhai Could you kindly change this and let me know? So that I can review again?

**Comment by neonbhai** [comment]
@abdulrahuman5196 Please take a look at the changes 🙇

File: src/libs/migrations/CheckForPreviousReportActionID.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * @returns {Promise<void>}
+ */
+export default function () {
+    return getReportActionsFromOnyx().then((allReportActions) => {
+        if (!isValid(allReportActions)) {
+            Log.info(`[Migrate Onyx] Skipped migration CheckForPreviousReportActionID because there were no reportActions`);
+            return;
+        }
+
+        const firstValidReportAction = _.find(_.values(allReportActions), (reportAction) => isValid(reportAction));
```
Comment (abdulrahuman5196) [inline src/libs/migrations/CheckForPreviousReportActionID.js:45]: I don't think we need a new isValid function, we can use the lodashHas to find a report action with a `reportActionID` field. Then We can use the reportAction to find flow to the next migration step.

@neonbhai

**Review by abdulrahuman5196 — CHANGES_REQUESTED** [review]

**Comment by abdulrahuman5196** [comment]
And Just to note since this PR has already been delayed. I will take this PR my first priority now. And will be able to review anytime today. @neonbhai

File: src/libs/migrations/CheckForPreviousReportActionID.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ * @returns {Promise<void>}
+ */
+export default function () {
+    return getReportActionsFromOnyx().then((allReportActions) => {
+        if (_.isEmpty(allReportActions)) {
+            Log.info(`[Migrate Onyx] Skipped migration CheckForPreviousReportActionID because there were no reportActions`);
+            return;
+        }
+
+        const firstValidReportAction = _.find(_.values(allReportActions), (reportActions) => !_.isEmpty(reportActions));
```
Comment (abdulrahuman5196) [inline src/libs/migrations/CheckForPreviousReportActionID.js:35]: `!_.isEmpty(reportActions)` doesn't mean it will have a valid report action @neonbhai 

What if the input is like this, this case it will fail to identify?
you could add a inner _.find here or basically run `for` loop on `_.values(allReportActions)` and find the `firstValidValue`

```
r1: {
     1: {
              message: 1,
       },
       2: {
               message: 2,
       },
},
r2: {
    1: {
        reportActionID: 1,
     }
}

```

Comment (neonbhai) [reply src/libs/migrations/CheckForPreviousReportActionID.js:35]: Was assuming all `reportActions` would have a `reportActionID` key and if we have some keys, that would indicate the reportAction loaded correctly?

If not,

Since we cannot get the exact `reportActionData` that passes the inner check, as the outer find returns the first passing `reportActions` object. To get the first valid value we would have to execute a find again. Would this be satisfactory?

```javascript
const firstValidReportAction = _.find(_.values(allReportActions), (reportActions) => _.find(_.values(reportActions), (reportActionData) => _.has(reportActionData, 'reportActionID')));

const firstValidValue = _.find(_.values(firstValidReportAction), (reportActionData) => _.has(reportActionData, 'reportActionID'));
```

Comment (abdulrahuman5196) [reply src/libs/migrations/CheckForPreviousReportActionID.js:35]: >To get the first valid value we would have to execute a find again. Would this be satisfactory?
 
 Or we can basically run for loop on _.values(allReportActions) as suggested?

>do you think it would be safe to assume if we have some keys, that would mean reportAction integrity is maintained

We are making sure we don't miss any edge cases in case of zoombie reportActions. Since reportActionID is a mandatory key it should better to check if its present or not.

Comment (abdulrahuman5196) [reply src/libs/migrations/CheckForPreviousReportActionID.js:35]: @neonbhai ^

Comment (neonbhai) [reply src/libs/migrations/CheckForPreviousReportActionID.js]: @abdulrahuman5196 Added these. Please check the changes :bow:

File: tests/unit/MigrationTest.js (hunk 1 above)
Comment (abdulrahuman5196) [reply tests/unit/MigrationTest.js:645]: Can we remove this? Even if we change the order there is no change in result?

File: src/libs/migrations/CheckForPreviousReportActionID.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            callback: (allReportActions) => {
+                Onyx.disconnect(connectionID);
+                return resolve(allReportActions);
+            },
+        });
+    });
+}
+
+/**
+ * Migrate Onyx data for reportActions. If the first reportAction of a reportActionsForReport
```
Comment (abdulrahuman5196) [inline src/libs/migrations/CheckForPreviousReportActionID.js:23]:
```suggestion
 * Migrate Onyx data for reportActions. If the first valid reportAction of a report in onyx 
```

File: src/libs/migrations/CheckForPreviousReportActionID.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            Log.info(`[Migrate Onyx] CheckForPreviousReportActionID Migration: previousReportActionID found. Migration complete`);
+            return;
+        }
+
+        // If previousReportActionID not found:
+        Log.info(`[Migrate Onyx] CheckForPreviousReportActionID Migration: removing all reportActions because previousReportActionID not found in the first valid reportAction`);
+
+        const onyxData = {};
+        _.each(allReportActions, (reportAction, onyxKey) => {
+            if (_.isEmpty(reportAction)) {
```
Comment (abdulrahuman5196) [inline src/libs/migrations/CheckForPreviousReportActionID.js:62]: We don't need this check. It will leave inconsistent report actions in onyx.
Like for `[r1]: null` the reportAction will have `null` value while others have `{}`. If we remove this check, it will consistently have `{}` across all

**Review by abdulrahuman5196 — CHANGES_REQUESTED** [review]
@neonbhai Just some comments. Other than that all looks fine to me.

File: src/libs/migrations/CheckForPreviousReportActionID.js (hunk 7 above)
Comment (neonbhai) [reply src/libs/migrations/CheckForPreviousReportActionID.js]: @abdulrahuman5196 done

File: tests/unit/MigrationTest.js (hunk 2 above)
Comment (neonbhai) [reply tests/unit/MigrationTest.js]: all done!

File: tests/unit/MigrationTest.js (hunk 1 above)
Comment (neonbhai) [reply tests/unit/MigrationTest.js]: done

**Comment by abdulrahuman5196** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>




https://github.com/Expensify/App/assets/46707890/42679c42-7f9b-4e5d-acf4-259e8b1216ef




</details>

<details>
<summary>Mobile Web - Chrome</summary>






</details>

<details>
<summary>Mobile Web - Safari</summary>







</details>

<details>
<summary>Desktop</summary>







</details>

<details>
<summary>iOS</summary>






</details>

<details>
<summary>Android</summary>






</details>

**Review by abdulrahuman5196 — APPROVED** [review]
Changes looks good and works well. Reviewers checklist is also complete.

All yours. @roryabraham 

🎀 👀 🎀
C+ Reviewed

**Review by roryabraham — APPROVED** [review]
LGTM 👍🏼

**Comment by mvtglobally** [comment]
@neonbhai @abdulrahuman5196 @roryabraham 

Anything we need to QA in staging? 
Can you pls update steps if yes
